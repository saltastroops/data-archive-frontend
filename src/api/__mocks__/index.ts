const mockBaseAxiosClient = jest.genMockFromModule("..");

const baseAxiosClient: any = ((mockBaseAxiosClient as any).mockBaseAxiosClient = jest.fn());

export { baseAxiosClient };
