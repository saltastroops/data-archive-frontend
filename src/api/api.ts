import { baseAxiosClient } from "./";

/**
 * An API for the Data Archive.
 *
 * login
 *   A function responsible for authenticating the user.
 */
export default {
  /**
   * Log the user in.
   *
   * Parameters:
   * -----------
   * credentials:
   *     The authentication credentials. They must contain the username and
   *     password.
   *
   * Returns:
   * --------
   * true
   */
  login: async (credentials: { password: string; username: string }) => {
    return baseAxiosClient().post("/auth/login", {
      password: credentials.password,
      username: credentials.username
    });
  },

  /**
   * Log the user out.
   *
   * Returns:
   * --------
   * true
   */
  logout: async () => {
    return baseAxiosClient().post("/auth/logout", {});
  }
};
