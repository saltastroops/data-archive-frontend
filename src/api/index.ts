import axios, { AxiosRequestConfig } from "axios";

/**
 * An axios base client responsible for the HTTP requests.
 *
 * If any of the given configuration parameters clashes with a default one, the
 * given value is used.
 *
 * Parameters
 * ----------
 * config: object
 *     Configuration parameters for the Axios client.
 */
export const baseAxiosClient = (config?: AxiosRequestConfig) => ({
  _client: axios.create(configuration(config)),

  _handleError(error: any) {
    if (error.response && error.response.data && error.response.data.message) {
      throw new Error(error.response.data.message);
    } else {
      throw error;
    }
  },

  get(url: string, conf: AxiosRequestConfig = {}) {
    return this._client.get(url, conf).catch(this._handleError);
  },

  post(url: string, data: any) {
    return this._client.post(url, data).catch(this._handleError);
  },
});

/**
 * Configuration parameters for the Axios client. If the given configuration
 * contains fields of the default configuration, the values of the custom
 * configuration are used.
 *
 * Parameters
 * ----------
 * config: object
 *     Axios configuration.
 */
function configuration(config?: { [key: string]: any }) {
  const defaultConfig = {
    ...config,
    baseURL: `${process.env.REACT_APP_BACKEND_URI}`,
    withCredentials: true,
  };

  return Object.assign({}, defaultConfig, config || {});
}
