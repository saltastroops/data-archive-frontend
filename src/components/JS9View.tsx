import noScroll from "no-scroll";
import React from "react";
import CSSTransition from "react-transition-group/CSSTransition";
import styled from "styled-components";
import { baseAxiosClient } from "../api";

const Overlay = styled.div`
  background: rgba(0, 0, 0, 0.75);
  display: flex;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 50;
  padding: 1.2rem;
`;

const Modal = styled.div`
  max-width: 800px;
  position: relative;
  padding: 1.2rem;
  background: #ffffff;
  background-clip: padding-box;
  box-shadow: 0 12px 15px 0 rgba(0, 0, 0, 0.25);
  margin: auto;
`;

const CloseButton = styled.button`
  position: absolute;
  top: 14px;
  right: 14px;
  border: none;
  padding: 0;
  background-color: transparent;
  display: flex;
`;

interface IJS9ViewProps {
  closeIconSize?: number;
  fitsURL: string;
  onClose: () => void;
  open: boolean;
}

interface IJS9ViewState {
  open: boolean;
}

/**
 * A modal for previewing FITS files with JS9.
 *
 * When using this component, keep the following in mind:
 *
 * - This component may only be used once. (More technically, it must only be
 *   mounted once.
 * - The component should not be included within a router component or any
 *   other component seriously meddling with the DOM.
 * - The component must be initialised with an empty string for its fitsURL
 *   property.
 *
 * You should update the component with the help of the JS9ViewContext context.
 *
 * The modal part of this component is essentially a stripped-down version of
 * react-responsive=-model
 * (https://www.npmjs.com/package/react-responsive-modal).
 */
class JS9View extends React.Component<IJS9ViewProps, IJS9ViewState> {
  static initialisedBefore = false;

  static blockScroll() {
    noScroll.on();
  }

  shouldClose: boolean | null = null;

  close = () => {
    // Close pop up windows sich as the one for displaying the header
    document
      .querySelectorAll("#dhtmlwindowholder .dhtmlwindow")
      .forEach((e) => ((e as HTMLElement).style.display = "none"));

    this.props.onClose();
  };

  handleOpen = () => {
    document.addEventListener("keydown", this.handleKeydown);
  };

  handleClose = () => {
    document.removeEventListener("keydown", this.handleKeydown);
  };

  handleClickOverlay = () => {
    if (this.shouldClose === null) {
      this.shouldClose = true;
    }

    if (!this.shouldClose) {
      this.shouldClose = null;
      return;
    }

    this.close();

    this.shouldClose = null;
  };

  handleClickCloseIcon = (event: any) => {
    this.close();
  };

  handleKeydown = (event: any) => {
    if (event.keyCode !== 27) {
      return;
    }

    this.close();
  };

  handleModalEvent = () => {
    this.shouldClose = false;
  };

  componentDidMount() {
    if (JS9View.initialisedBefore) {
      throw new Error("The JS9View component must only be used once.");
    }

    if (this.props.fitsURL) {
      throw new Error(
        "The JS9View component must be initialised with an empty string for the fitsURL property."
      );
    }

    if ((window as any).JS9) {
      (window as any).JS9.init();
    }

    window.addEventListener("keydown", this.handleKeydown);

    JS9View.initialisedBefore = true;
  }

  componentWillUnmount() {
    window.removeEventListener("keydown", this.handleKeydown);
  }

  async componentDidUpdate(
    prevProps: Readonly<IJS9ViewProps>,
    prevState: Readonly<IJS9ViewState>,
    snapshot?: any
  ): Promise<void> {
    const { fitsURL } = this.props;

    if (fitsURL && fitsURL !== prevProps.fitsURL) {
      if ((window as any).JS9) {
        try {
          const fits = await baseAxiosClient({ responseType: "blob" }).get(
            fitsURL
          );
          (window as any).JS9.Load(fits.data);
        } catch (e) {
          alert(e.message);
        }
      }
    }
  }

  render() {
    const { closeIconSize, open } = this.props;

    return (
      <div data-test="wrapper" style={{ display: open ? "block" : "none" }}>
        <CSSTransition in={open} appear={true} classNames={{}} timeout={500}>
          <Overlay
            data-test="overlay"
            onKeyDown={this.handleKeydown}
            onClick={this.handleClickOverlay}
          >
            <Modal
              onMouseDown={this.handleModalEvent}
              onMouseUp={this.handleModalEvent}
              onClick={this.handleModalEvent}
              role="dialog"
              aria-modal="true"
            >
              <div className="section">
                <div className="JS9Menubar" />
                <div className="JS9" />
                <div className="JS9Colorbar" />
              </div>
              <CloseButton data-test="close" onClick={this.close}>
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  width={closeIconSize || 28}
                  height={closeIconSize || 28}
                  viewBox="0 0 36 36"
                >
                  <path d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z" />
                </svg>
              </CloseButton>
            </Modal>
          </Overlay>
        </CSSTransition>
      </div>
    );
  }
}

export default JS9View;

/**
 * A context for updating the JS9View component. It includes the following
 * properties:
 *
 * - close: Function for closing the  modal.
 * - load: Function for loading a FITS image.
 * - open: Function for opening the modal.
 *
 * A call to the load function should not automatically open the modal.
 */
export const JS9ViewContext = React.createContext({
  close: () => {
    // do nothing
  },
  load: (fitsURL: string) => {
    // do nothing
  },
  open: () => {
    // do nothing
  },
});
