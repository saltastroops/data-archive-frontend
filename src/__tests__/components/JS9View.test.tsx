jest.mock("../../api");

import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { baseAxiosClient } from "../../api";
import JS9View from "../../components/JS9View";

describe("JS9View", () => {
  (window as any).JS9 = {
    Load: jest.fn(),
    init: jest.fn(),
  };

  const onClose = jest.fn();

  afterEach(() => {
    (window as any).JS9.init.mockReset();
    (window as any).JS9.Load.mockReset();
    onClose.mockReset();
    JS9View.initialisedBefore = false;
  });

  it("should display correctly for open=true", () => {
    const wrapper = mount(<JS9View open={true} fitsURL="" onClose={onClose} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should display correctly for open=false", () => {
    const wrapper = mount(
      <JS9View open={false} fitsURL="" onClose={onClose} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should throw an error if used more than once", () => {
    const f = () =>
      mount(
        <div>
          <JS9View open={false} fitsURL="" onClose={onClose} />
          <JS9View open={false} fitsURL="" onClose={onClose} />
        </div>
      );

    // suppress error messages
    const originalError = console.error;
    console.error = jest.fn();

    expect(f).toThrowError();

    // re-enable error messages
    console.error = originalError;
  });

  it("should throw an error if initialised with a truthy FITS URL", () => {
    const f = () =>
      mount(
        <div>
          <JS9View
            open={false}
            fitsURL="http://some.url.saao.ac.za"
            onClose={onClose}
          />
        </div>
      );

    // suppress error messages
    const originalError = console.error;
    console.error = jest.fn();

    expect(f).toThrowError();

    // re-enable error messages
    console.error = originalError;
  });

  it("should initialize JS9", () => {
    const wrapper = mount(
      <JS9View open={false} fitsURL="" onClose={onClose} />
    );

    expect((window as any).JS9.init).toHaveBeenCalled();
  });

  it("should load a new FITS URL if it is not an empty string", () => {
    const url1 = "http://some.url.saao.ac.za";
    const url2 = "http://some.other.url.saao.ac.za";

    (baseAxiosClient as any).mockImplementation(() => ({
      get: () => ({ data: "image" }),
    }));

    (window as any).JS9.Load.mockReturnValue("image");

    const wrapper = mount(
      <JS9View open={false} fitsURL="" onClose={onClose} />
    );

    try {
      wrapper.setProps({ fitsURL: url1 });
      wrapper.setProps({ fitsURL: url2 });

      expect((window as any).JS9.Load).toHaveBeenCalledTimes(2);
      expect((window as any).JS9.Load).toHaveBeenNthCalledWith(1, url1);
      expect((window as any).JS9.Load).toHaveBeenNthCalledWith(2, url2);
    } catch (e) {
      return;
    }
  });

  it("should not load a new FITS URL if it is an empty string", () => {
    const url1 = "http://some.url.saao.ac.za";

    const wrapper = mount(
      <JS9View open={false} fitsURL="" onClose={onClose} />
    );

    (baseAxiosClient as any).mockImplementation(() => ({
      get: () => ({ data: "image" }),
    }));

    (window as any).JS9.Load.mockReturnValue("image");

    try {
      wrapper.setProps({ fitsURL: url1 });
      wrapper.setProps({ fitsURL: "" });

      expect((window as any).JS9.Load).toHaveBeenCalledTimes(1);
      expect((window as any).JS9.Load).toHaveBeenCalledWith(url1);
    } catch (e) {
      return;
    }
  });

  it("should not load a FITS URL if another property has changed", () => {
    const url1 = "http://some.url.saao.ac.za";

    (baseAxiosClient as any).mockImplementation(() => ({
      get: () => ({ data: "image" }),
    }));

    (window as any).JS9.Load.mockReturnValue("image");

    const wrapper = mount(
      <JS9View open={false} fitsURL="" onClose={onClose} />
    );

    try {
      wrapper.setProps({ fitsURL: "http://some.url.saao.ac.za" });
      expect((window as any).JS9.Load).toHaveBeenCalled();
    } catch (e) {
      return;
    }

    (window as any).JS9.Load.mockReset();

    try {
      wrapper.setProps({ open: true });
      expect((window as any).JS9.Load).not.toHaveBeenCalled();
    } catch (e) {
      return;
    }
  });

  it("should call onClose when the overlay is clicked", () => {
    const wrapper = mount(<JS9View open={true} fitsURL="" onClose={onClose} />);

    // Simulate a click on the overlay
    wrapper.find('div[data-test="overlay"]').simulate("click");

    // The onClose property has been called
    expect(onClose).toHaveBeenCalled();
  });

  it("should close when the close button is clicked", () => {
    const wrapper = mount(<JS9View open={true} fitsURL="" onClose={onClose} />);

    // Simulate a click on the close button
    wrapper.find('button[data-test="close"]').simulate("click");

    // The onClose property has been called
    expect(onClose).toHaveBeenCalled();
  });

  it("should close when the esc key is clicked", () => {
    const wrapper = mount(<JS9View open={true} fitsURL="" onClose={onClose} />);

    // Simulate using the esc key
    wrapper
      .find('div[data-test="overlay"]')
      .simulate("keydown", { keyCode: 27 });

    // The onClose property has been called
    expect(onClose).toHaveBeenCalled();
  });

  it("should not close when a key other than esc is clicked", () => {
    const wrapper = mount(<JS9View open={true} fitsURL="" onClose={onClose} />);

    // Simulate using a key other than esc
    wrapper
      .find('div[data-test="overlay"]')
      .simulate("keydown", { keyCode: 27 });

    // The onClose property has been called
    expect(onClose).toHaveBeenCalled();
  });
});
