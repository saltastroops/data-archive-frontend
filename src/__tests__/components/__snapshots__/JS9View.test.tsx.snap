// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS9View should display correctly for open=false 1`] = `
.c0 {
  background: rgba(0,0,0,0.75);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 50;
  padding: 1.2rem;
}

.c1 {
  max-width: 800px;
  position: relative;
  padding: 1.2rem;
  background: #ffffff;
  background-clip: padding-box;
  box-shadow: 0 12px 15px 0 rgba(0,0,0,0.25);
  margin: auto;
}

.c2 {
  position: absolute;
  top: 14px;
  right: 14px;
  border: none;
  padding: 0;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<JS9View
  fitsURL=""
  onClose={[MockFunction]}
  open={false}
>
  <div
    data-test="wrapper"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <CSSTransition
      appear={true}
      classNames={Object {}}
      in={false}
      timeout={500}
    >
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={500}
        unmountOnExit={false}
      >
        <styled.div
          data-test="overlay"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <StyledComponent
            data-test="overlay"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": false,
                  "lastClassName": "c0",
                  "rules": Array [
                    "
  background: rgba(0, 0, 0, 0.75);
  display: flex;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 50;
  padding: 1.2rem;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <div
              className="c0"
              data-test="overlay"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
              <styled.div
                aria-modal="true"
                onClick={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                role="dialog"
              >
                <StyledComponent
                  aria-modal="true"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": false,
                        "lastClassName": "c1",
                        "rules": Array [
                          "
  max-width: 800px;
  position: relative;
  padding: 1.2rem;
  background: #ffffff;
  background-clip: padding-box;
  box-shadow: 0 12px 15px 0 rgba(0, 0, 0, 0.25);
  margin: auto;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  role="dialog"
                >
                  <div
                    aria-modal="true"
                    className="c1"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="dialog"
                  >
                    <div
                      className="section"
                    >
                      <div
                        className="JS9Menubar"
                      />
                      <div
                        className="JS9"
                      />
                      <div
                        className="JS9Colorbar"
                      />
                    </div>
                    <styled.button
                      data-test="close"
                      onClick={[Function]}
                    >
                      <StyledComponent
                        data-test="close"
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-htpNat",
                              "isStatic": false,
                              "lastClassName": "c2",
                              "rules": Array [
                                "
  position: absolute;
  top: 14px;
  right: 14px;
  border: none;
  padding: 0;
  background-color: transparent;
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.button",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-htpNat",
                            "target": "button",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                      >
                        <button
                          className="c2"
                          data-test="close"
                          onClick={[Function]}
                        >
                          <svg
                            height={28}
                            viewBox="0 0 36 36"
                            width={28}
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
                            />
                          </svg>
                        </button>
                      </StyledComponent>
                    </styled.button>
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </StyledComponent>
        </styled.div>
      </Transition>
    </CSSTransition>
  </div>
</JS9View>
`;

exports[`JS9View should display correctly for open=true 1`] = `
.c0 {
  background: rgba(0,0,0,0.75);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 50;
  padding: 1.2rem;
}

.c1 {
  max-width: 800px;
  position: relative;
  padding: 1.2rem;
  background: #ffffff;
  background-clip: padding-box;
  box-shadow: 0 12px 15px 0 rgba(0,0,0,0.25);
  margin: auto;
}

.c2 {
  position: absolute;
  top: 14px;
  right: 14px;
  border: none;
  padding: 0;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<JS9View
  fitsURL=""
  onClose={[MockFunction]}
  open={true}
>
  <div
    data-test="wrapper"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <CSSTransition
      appear={true}
      classNames={Object {}}
      in={true}
      timeout={500}
    >
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={500}
        unmountOnExit={false}
      >
        <styled.div
          data-test="overlay"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <StyledComponent
            data-test="overlay"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": false,
                  "lastClassName": "c0",
                  "rules": Array [
                    "
  background: rgba(0, 0, 0, 0.75);
  display: flex;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow-y: auto;
  overflow-x: hidden;
  z-index: 50;
  padding: 1.2rem;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <div
              className="c0"
              data-test="overlay"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
              <styled.div
                aria-modal="true"
                onClick={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                role="dialog"
              >
                <StyledComponent
                  aria-modal="true"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": false,
                        "lastClassName": "c1",
                        "rules": Array [
                          "
  max-width: 800px;
  position: relative;
  padding: 1.2rem;
  background: #ffffff;
  background-clip: padding-box;
  box-shadow: 0 12px 15px 0 rgba(0, 0, 0, 0.25);
  margin: auto;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  role="dialog"
                >
                  <div
                    aria-modal="true"
                    className="c1"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="dialog"
                  >
                    <div
                      className="section"
                    >
                      <div
                        className="JS9Menubar"
                      />
                      <div
                        className="JS9"
                      />
                      <div
                        className="JS9Colorbar"
                      />
                    </div>
                    <styled.button
                      data-test="close"
                      onClick={[Function]}
                    >
                      <StyledComponent
                        data-test="close"
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-htpNat",
                              "isStatic": false,
                              "lastClassName": "c2",
                              "rules": Array [
                                "
  position: absolute;
  top: 14px;
  right: 14px;
  border: none;
  padding: 0;
  background-color: transparent;
  display: flex;
",
                              ],
                            },
                            "displayName": "styled.button",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-htpNat",
                            "target": "button",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onClick={[Function]}
                      >
                        <button
                          className="c2"
                          data-test="close"
                          onClick={[Function]}
                        >
                          <svg
                            height={28}
                            viewBox="0 0 36 36"
                            width={28}
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
                            />
                          </svg>
                        </button>
                      </StyledComponent>
                    </styled.button>
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </StyledComponent>
        </styled.div>
      </Transition>
    </CSSTransition>
  </div>
</JS9View>
`;
